### PR TITLE
Eslint config updates and ESM migration

### PR DESCRIPTION
**Pull request by flurdy**
- Migrate the codebase to `ESM`
- Update `Eslint` config
- Fix issues brought up by the new `Eslint` config
- Fix type issues
- Stop tests from hanging
- Update `e2e` tests
- Update docs
  - Add testing docs to cover:
    - Unit tests
    - Smoke Tests
    - E2e tests 

## Checks performed

- `cdp-uploader-smoke-tests` ✅
- `npm test` ✅
- `npm run test:e2e` ✅
- `npm run lint` ✅
- Docker build and run ✅

### CODE REVIEW

**Comment by flurdy** [comment]
Will fix journey test https://github.com/DEFRA/cdp-uploader/actions/runs/10376556225/job/28728780564?pr=106#step:7:21

File: src/server/common/helpers/redis/redis-client.js [hunk 1]
```diff
@@ -1,45 +1,68 @@
...
-function buildRedisClient(redisConfig) {
+export function buildRedisClient(redisConfig) {
   const logger = createLogger()
   const port = 6379
   const db = 0
   const keyPrefix = redisConfig.keyPrefix
   const host = redisConfig.host
+  let redisClient
 
-  const client = redisConfig.useSingleInstanceCache
```
Comment (chotai) [inline src/server/common/helpers/redis/redis-client.js:20]: Think this should be:
```  
const client = redisConfig.useSingleInstanceCache
    ? new Redis({ port, host, db, keyPrefix })
    : new Cluster([{ host, port }], {
        keyPrefix,
        slotsRefreshTimeout: 10000,
        slotsRefreshTimeout: 2000,
        dnsLookup: (address, callback) => callback(null, address),
        redisOptions: {
          username: redisConfig.username,
          password: redisConfig.password,
          db,
          tls: {}
        }
```
There's a PR on the frontend we need to merge

**Review by chotai — APPROVED** [review]

File: package.json [hunk 2]
```diff
@@ -61,48 +66,66 @@
...
+    "@types/jest": "29.5.12",
+    "@types/lodash": "4.17.7",
+    "@types/node": "20.14.12",
+    "@types/node-fetch": "2.6.11",
+    "@types/nunjucks": "3.2.6",
+    "@types/webpack-assets-manifest": "5.1.4",
+    "@typescript-eslint/eslint-plugin": "7.16.0",
+    "@typescript-eslint/parser": "7.17.0",
+    "autoprefixer": "10.4.19",
     "babel-jest": "29.7.0",
```
[the remark this answers is not in the document]
Comment (flurdy) [reply package.json:91]: Babel is used in one of test bits still I think

File: src/server/common/helpers/redis/redis-client.js (hunk 1 above)
Comment (flurdy) [reply src/server/common/helpers/redis/redis-client.js:20]: Yeah but that is separate to the ESM change?

File: package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (feedmypixel) [reply package.json:91]: Yeah babel is still used in tests

File: src/server/common/helpers/redis/redis-client.js (hunk 1 above)
Comment (feedmypixel) [reply src/server/common/helpers/redis/redis-client.js:20]: Can we pick this up as a separate thing? As there is so much already in here anything extra would be good done in a different PR

File: package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (flurdy) [reply package.json:91]: (as in jest)

File: babel.config.cjs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+/**
+ * @type {TransformOptions}
+ */
+module.exports = {
+  browserslistEnv: 'node',
+  presets: [
+    [
+      '@babel/preset-env',
+      {
+        modules: NODE_ENV === 'test' ? 'auto' : false
```
[the remark this answers is not in the document]
Comment (feedmypixel) [reply babel.config.cjs:12]: The `@babel/preset-env` is using the value defined in `.browserslistrc` which is node 20

File: .eslintrc.cjs [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        ],
+        'no-console': 'error',
+
+        // Turn off strict type checking rules
+        '@typescript-eslint/no-unsafe-argument': 'off',
+        '@typescript-eslint/no-unsafe-assignment': 'off',
+        '@typescript-eslint/no-unsafe-call': 'off',
+        '@typescript-eslint/no-unsafe-member-access': 'off',
+        '@typescript-eslint/no-unsafe-return': 'off',
+
```
[the remark this answers is not in the document]
Comment (feedmypixel) [reply .eslintrc.cjs:52]: Its set further down for tests. FYI this is the config taken from the `cdp-node-frontend-template` we can add if it you like? If so would be good to get in to the template also.

File: src/server/common/helpers/redis/redis-client.js (hunk 1 above)
Comment (flurdy) [reply src/server/common/helpers/redis/redis-client.js:20]: That is basically what the current code does atm, just slightly different conditional branching. Outcome is the same

**Pull request merged**